src/abstractdataframe/join.jl

```diff
@@ -247,6 +248,13 @@ Join two `DataFrame` objects
   if `true`, duplicate names will be suffixed with `_i`
   (`i` starting at 1 for the first duplicate).
 
+* `indicator` : Default: `false`. If `true`, adds categorical indicator
```

Comment:
I thought we agreed at #1412 that it would be better to have two separate keywords allowing the user the specify the name of the columns to create, with a default to `nothing`.

Reply:
I thought we agreed we didn't want a column of booleans, but I took the thumbs up on my last post suggesting this categorical solution (same as pandas and stata) as confirmation that made sense. I can revert them to separate bit columns, but I feel like users will usually want the interaction of the two, jumping to a categorical with all cases seemed logical to me.

Reply:
I'm not sure what's the most practical for users. Do you have examples of operations people may want to do with these columns? It sounds simpler to me to do `df[:left]` than `df[:merge] .== "left_only"` (and it's also more efficient and takes less memory).

Reply:
So `df[:left]` and `df[:merge] .== "left_only"` are different operations. The first will get you all observations that were in `df1`. The second will get you observations that were in `df1` but not in `df2`. 

Personally, I use this a lot for doing things like `@assert nrow(df[df[:_merge] .== "left_only"]) == 0` to validate that I successfully found a match for all keys in the left dataset in the right dataset.

Reply:
OK. `nrow(df[df[:_merge] .== "left_only"]) == 0` can be written more simply as `!any(df[:_merge] .== "left_only"])` or `!("left_only" in df[:_merge])`. With two Boolean columns it would be `!any(df[:left] .& .!df[:right])`. Personally I find the latter a bit cleaner (no need to remember whether it"s `"left_only"`, `"left only"`, `"only left"` or any other variation), but while it should be faster it's also slightly more convoluted.

But since you say you use it as a validation check, aren't there better ways of achieving this? For example, doing an inner join and then checking `nrow(joined) == nrow(df1)`? Or maybe we should add a keyword argument to do that?

Reply:
> But since you say you use it as a validation check, aren't there better ways of achieving this? For example, doing an inner join and then checking nrow(joined) == nrow(df1)? Or maybe we should add a keyword argument to do that?

The indicator column gives you both sides immediately, and only requires one merge (nice computationally). For example, say I think the data should have all keys match (so `outer` should be same as `inner`), but I want to be sure everything works, so I do an `outer` join. I could then do a `right` join (to check for unmatched `df2` keys) and a `left` join (to check for unmatched `df1` keys) and compare all of those. But that's three merges. Here it's one column from one merge. 

It's also the case that I often leave those simple tests in place in my code for defensive purposes, but when a problem comes up I want to query those bad matches, for which `nrow` is not enough. So if my test fails, I'd want to look at the `df[:_merge] .== "left_only"]`. 

And yes, I'm of the view that `df[:_merge] .== "left_only"]` is more readable / like english than `df[df[:_left] .& .!df[:_right]` to me.

Reply:
> The indicator column gives you both sides immediately, and only requires one merge (nice computationally). For example, say I think the data should have all keys match (so outer should be same as inner), but I want to be sure everything works, so I do an outer join. I could then do a right join (to check for unmatched df2 keys) and a left join (to check for unmatched df1 keys) and compare all of those. But that's three merges. Here it's one column from one merge.

My example also only requires one merge. If you do an inner join and get as many columns as `df1`, that means all rows matched.

Anyway I'm not saying we shouldn't have the ability to create that new column, but maybe we also want a convenience argument to ensure all rows matched?

@bkamins Opinions?

Reply:
> My example also only requires one merge. If you do an inner join and get as many columns as df1, that means all rows matched.

That's true, sorry -- was thinking ahead to the case of wanting to see which observations failed if it didn't work.

Reply:
As I have stated in the issue. I would prefer two columns of boolean indicators as this will be a faster and more memory efficient operation.

But @nickeubank knows his usecase best, and if one column is much more convenient for him I would be OK with this. The only question is if we want to use `String` or `Symbol` (I guess `Symbol` like `:left_only`, `:right_only` and `:both` should be a bit faster and less memory hungry as `Symbol`s are interned - this would have to be tested).

Finally I have a strong preference for a calling convention. Instead of `indicator` being `Bool` I would like it to be `Union{Nothing, Symbol}` and `nothing` being default. In such a situation if some user wants `indicator` column she would pass the column name that suits her best. Then if duplicate column name is encountered it would be handled following value of `makeunique`.

(the same would be recommended if we went for two columns of `Bool`)

Reply:
> But @nickeubank knows his usecase best, and if one column is much more convenient for him I would be OK with this. The only question is if we want to use String or Symbol (I guess Symbol like :left_only, :right_only and :both should be a bit faster and less memory hungry as Symbols are interned - this would have to be tested).

We should definitely use a `CategoricalArray` for that, it would be much more efficient (with a `UInt8` reference type it would use 8× less memory than a 64-bit pointer). Then a `String` (actually`CategoricalString`) element type sounds the most natural.

BTW, `"right"` and `"left"` sound better than `"right_only"` and `"left_only"`, don't they?

> Finally I have a strong preference for a calling convention. Instead of indicator being Bool I would like it to be Union{Nothing, Symbol} and nothing being default. In such a situation if some user wants indicator column she would pass the column name that suits her best. Then if duplicate column name is encountered it would be handled following value of makeunique.

Agreed.

Reply:
`CategoricalArray` is indeed most natural

Reply:
> Finally I have a strong preference for a calling convention. Instead of indicator being Bool I would like it to be Union{Nothing, Symbol} and nothing being default. In such a situation if some user wants indicator column she would pass the column name that suits her best. Then if duplicate column name is encountered it would be handled following value of makeunique.

Happy to do that. Will update. 

> BTW, "right" and "left" sound better than "right_only" and "left_only", don't they?

I think `right_only` is more descriptively accurate. All observations in `both` are in the right dataset, and also in the left dataset. The point is the observations labeled `right_only` are, in fact, only in the right dataset. 

Also one more thought on an earlier point: the other limitation to checking using something like  `nrow(joined) == nrow(df1)` is that it only works for a 1:1 merge. If the merge is 1:many, that test becomes uninformative. :/